pkg/runner/common_result.go

```diff
@@ -10,13 +11,49 @@ type Result struct {
 	Journal  *Journal                 `json:"journal"`
 }
 
-func newResult() *Result {
-	return &Result{
+func newResult(input *api.RunInput) *Result {
+	result := &Result{
 		Outcome:  task.OutcomeUnknown,
 		Outcomes: make(map[string]*GroupOutcome),
 		Journal: &Journal{
 			Events:       make(map[string]string),
 			PodsStatuses: make(map[string]struct{}),
 		},
 	}
+
+	for _, g := range input.Groups {
+		result.Outcomes[g.ID] = &GroupOutcome{
+			Total: g.Instances,
+			Ok:    0,
+		}
+	}
+
+	return result
+}
+
+func (r *Result) addOutcome(groupID string, outcome task.Outcome) {
+	switch outcome {
+	case task.OutcomeSuccess:
+		r.Outcomes[groupID].Ok++
+	default:
+		// skip
+	}
+}
+func (r *Result) countTotalInstances() int {
+	count := 0
+	for _, g := range r.Outcomes {
+		count += g.Total
+	}
+	return count
+}
+
+// TODO: this should be a getter instead of a mutation
+func (r *Result) updateOutcome() {
```

Comment:
Is the idea here to get rid of `r.Outcome` field altogether and replace places that read it with `r.getOutcome` that iterates through `r.Outcomes` and returns either `task.OutcomeOk` or `task.OutcomeFailure`?

Reply:
Exactly,
before the whole thing was implemented in the runner itself: https://github.com/testground/testground/blob/bcdcfb9be20c0273beaf225b9da674aaf2e30f75/pkg/runner/local_docker.go#L198-L238

Step 1: move this into the `Result` abstraction, which means we can reuse this abstraction in other runners (and drop the [result module](https://github.com/testground/testground/blob/master/pkg/data/result.go) which is a workaround)
Step 2: fix the abstraction